src/v/kafka/server/group_manager.cc

```diff
@@ -614,6 +628,104 @@ group_manager::txn_offset_commit(txn_offset_commit_request&& r) {
     }
 }
 
+static cluster::begin_group_tx_reply make_begin_tx_reply(cluster::tx_errc ec) {
+    cluster::begin_group_tx_reply reply;
+    reply.ec = ec;
+    return reply;
+}
+
+ss::future<cluster::begin_group_tx_reply>
+group_manager::begin_tx(cluster::begin_group_tx_request&& r) {
+    if (!_catchup_lock.take_reader_lock()) {
+        vlog(klog.warn, "can't process a tx: coordinator_load_in_progress");
+        co_return make_begin_tx_reply(
+          cluster::tx_errc::coordinator_load_in_progress);
+    }
+
+    try {
+        auto error = validate_group_status(
+          r.ntp, r.group_id, offset_commit_api::key);
+        if (error != error_code::none) {
+            _catchup_lock.release_reader_lock();
```

Comment:
lock-guard for `_catchup_lock` will make a lot of the axiomatic complexity in this method a lot simpler.

Reply:
@senior7515 I had a thought about it but then I realized that I'm not sure about relying on RAII in the co_await / co_return context. Can a compiler call a destructor just after initialization and before the end of the **async** method? Since we never use the guard "down the stack" then there is no reason to pass it the continuation and it maybe freed immediately.

Reply:
@rystsov  so in seastar, we've always had that issue. effectively all async code has that issue really. The only workaround is to make some of these objects pointers. 

THe pattern i've used is this. assume a function called `future<> manage_resource()`

```

future<> manage_resource(){
    return do_with( create_temporary_lock_guard(this), [this]{
         return do_manage_resource();
   });
}

// private
future<> do_manage_resource() {
      co_await foo();
      ......
     co_return;
} 
```

if you look inside seaster, there is a lot of `parent_function()` and `do_actual_parent_function()` nomenclature to work around this problem.

Reply:
this is an example: 

https://github.com/vectorizedio/redpanda/blob/dev/src/v/storage/segment.h#L152-L155

Reply:
in fact, here is exactly the pattern, i'd use: 

https://github.com/vectorizedio/redpanda/blob/dev/src/v/storage/segment.cc#L248-L256